### DISCUSSION TITLE
GrapheneOS's Sandboxed Play services

### DISCUSSION THREAD

**Discussion by dimqua** [Divested-Mobile/DivestOS-Build#22] [Ideas] [answered]
I think it'd be useful to integrate GrapheneOS's patches that basically allow to run google play services as unprivileged apps. LOS devs refused to integrate them, but they were already ported by someone else, more info here: https://forum.xda-developers.com/t/guide-grapheneoss-sandboxed-play-services-in-your-rom.4340557/

I'm aware about microG, but unfortunately these are some apps that don't work with it correctly. GrapheneOS's patches provide a compromise solution allowing users to use such apps (most of them) under the separate work profile (or user account) without the need of installing GPlay services system-wide.

**Comment by SkewedZeppelin** [comment, marked as answer]
I don't care about Google apps.
I don't use them. I don't want them.
They are proprietary.

DivestOS does not support Google apps.
DivestOS will not include microG or the GrapheneOS' Play Services sandbox.
Anyone is welcome to fork DivestOS for such if they desire.

**Reply by Nairou** [reply]
I respect this answer, and I don't care about Google apps either. However I'm running into an increasing problem of not receiving notifications for most apps, as they rely on Google libraries for cloud messaging.

I wonder if it would be worth bundling that one component from microG?

Barring that, would you have any suggestion for getting it installed after the fact? Most instructions I see require TWRP, which DivestOS doesn't use.

**Reply by SkewedZeppelin** [reply]
There is no single component of microG you can include for just notifications. The whole thing would need to be included, and including microG requires signature spoofing which is a security risk.

DivestOS has compile time support for including microG, and when done that way will enable a restricted signature spoofing patch.
Adding microG to DivestOS via TWRP or Magisk will break things.

**Reply by Nairou** [reply]
Thanks. I found the compile-time support in another post here, I may have to try building it.

I'm curious about signature spoofing being a security risk, have any references for that? I'd like to weigh the pros and cons for my own use.

**Reply by SkewedZeppelin** [reply]
The standard carried Signature Spoofing patch can be granted to any app. <--- this is where the biggest issue is
The version DivestOS has can only be granted to a system app signed with the same release keys.
The version CalyxOS has only grants to microG and microG alone.

**Reply by catdevnull** [reply]
If a patch is provided (possibly using something similar to GrapheneOS's method), it won't be merged? It's the only way to *safely* run apps that depend on Play Services. Otherwise, only problematic MicroG can be used. Sandboxed Play Services should be able to be used in work profile and other profiles specifically.

**Reply by SkewedZeppelin** [reply]
I don't care about apps that depend on Play Services.
If you want to cleanly port Sandboxed Play Services to 10/11/12 and backport any appropriate patches and continue to maintain it into the future as the upstream implementation evolves, then I will accept it.
It is far from a trivial task however.

**Reply by rsssxd** [reply]
If any developer wants to offer support for Play services, will DivestOS accept it?

**Reply by SkewedZeppelin** [reply]
@rsssxd See comment directly above.

**Comment by MarkCyberSecDev93** [comment]
@SkewedZeppelin
**Sandbox Google Play Services** is not a spyware and **FOSS** doesn't always solve the security issues.

Read:
https://wonderfall.dev/fdroid-issues/
and
https://blog.blueboxsec.org/post/the-illusion-of-open-source/

At least add some support for installing Sandbox Play Services. It doesn't need to be pre-installed like **GrapheneOS** did. But yeah, I'm not complaining just stating the facts and I'm not here to argue. 

The point of this comment is to (probably) change your mind. It's okay if you still do not agree everything I pointed out in this comment.

**Reply by SkewedZeppelin** [reply]
DivestOS exists to achieve a balance of privacy, security, and freedom on primarily old devices.
Play Services is a square and does not fit in that triangle.

**Comment by edern-appere** [comment]
Sorry to interrupt you guys but I have one (dumb) question (because I do not have installed DivestOs yet) : I thought i read somewhere that DivestOs came bundled with microG... is it the case or not ?

**Reply by SkewedZeppelin** [reply]
microG is NOT included or supported on DivestOS.

**Reply by edern-appere** [reply]
Ok, better to know it now. 
Is there a location provider installed so that some apps that require it are able to run "normaly" ?

**Reply by SkewedZeppelin** [reply]
UnifiedNlp was previously included, but it was removed due to concerns of it leaking location data to apps without location permission.

**Reply by edern-appere** [reply]
Is there an alternative to UnifiedNlp ? or some kind of workaround ? 
Would apps like Transportr still work without location provider ?

**Comment by edern-appere** [comment]
Well...  The only component I was using microg for is UnifiedNlp. I don't know if I'm able to deal with no location service provider. I have to check my apps...


Le 19 mai 2022 23:02:36 GMT+02:00, Tad ***@***.***> a écrit :

**Reply by SkewedZeppelin** [reply]
UnifiedNlp was removed due to concerns of it leaking location to apps without permission, among other reasons.

**Comment by michaelschefczyk** [comment]
MicroG is still listed on the DivestOS about page https://divestos.org/index.php?page=about

It would be super nice if everyone was clearly informed that it is not included. Any choice is fully understandable. However, there is also a user community looking for different compromises between privacy and location information quality.

**Reply by SkewedZeppelin** [reply]
That is the credits section, it is mentioned no where else such as the homepage, faq page, technical details page, post install page, or screenshots page.

(all pages which it was on at some point years ago)

There is even a note on the site [stating](https://divestos.org/index.php?page=broken#unsupported):
> Google Apps or microG or Sandboxed Play Services are NOT supported.

**Comment by Username23453** [comment]
@SkewedZeppelin , what about "gmaps wv", Its a google maps wrapper, does DivestOS support this?

**Reply by SkewedZeppelin** [reply]
GMaps WV is my app and doesn't use/require Play Services, it will work on any OS: https://divestos.org/index.php?page=our_apps#gmapswv

Actual Google Maps works on DivestOS too!

**Reply by Username23453** [reply]
@SkewedZeppelin ,thank you! sorry to bother. I appreciate it.

**Comment by SkewedZeppelin** [comment]
17.1 and higher have an unprivileged microG feature now: https://divestos.org/pages/faq#appCompatibility

**Reply by mortezadadgar** [reply]
Could you include same patches for 15.1 and up?

**Reply by SkewedZeppelin** [reply]
I started a backport for 16.0 but haven't finished it. Once that is done it should be easier to backport to 14 and 15.

**Comment by serafdev** [comment]
This is interesting, I'm not looking for Play Services but just a Sandboxing feature, I see in your faq this section:

https://divestos.org/pages/faq#sandbox
> How can I sandbox my apps?[¶](https://divestos.org/pages/faq#sandbox)
All apps on Android are already sandboxed. If you want to be able to grant [...] **files access** ...

I want to know if:

- Default Android does have files scoping
- If DivestOS has files scoping

If not is it possible to add that as files scoping should be part of sandboxing definition

@SkewedZeppelin thanks in advance!

**Reply by SkewedZeppelin** [reply]
DivestOS inherits the standard AOSP scoped storage permissions for apps.
https://source.android.com/docs/core/storage/scoped

It doesn't feature the more fine grained version that GrapheneOS has.
https://grapheneos.org/usage#storage-scopes

**Reply by serafdev** [reply]
Thanks @SkewedZeppelin 

I'm interested to knowing what needs to be done if we want to add storage and contact scopes as described in GrapheneOS

Is this something you are looking to add to the project? Is it something trivial to "merge" from the GrapheneOS repositories?

**Reply by SkewedZeppelin** [reply]
I'm not currently interested in porting it is it is quite complex.
Maybe I'll consider it after DivestOS stabilizes on A14 or A15.
